Allow translating dynamic variables in TypeScript with `as`/`satisfies`

ewlsh opened the pull request:
Adds support for `variable satisfies 'key'` and `variable as 'key'`

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

#### Checklist

- [ ] only relevant code is changed (make a diff before you submit the PR)
- [X] run tests `npm run test`
- [X] tests are included
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/i18next/.github/blob/master/CONTRIBUTING.md)

#### Checklist (for documentation change)

- [ ] only relevant documentation part is changed (make a diff before you submit the PR)
- [ ] motivation/reason is provided
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/i18next/.github/blob/master/CONTRIBUTING.md)

ewlsh commented:
#43 is an issue to track a discussion around this PR (this is marked as a draft intentionally)

adrai commented:
After further consideration and discussion (with a friend of mine), I've changed my position on this feature.

Initially, I was hesitant about adding `TsSatisfiesExpression` support directly to the core, thinking it might be better as a plugin. However, I now believe this is the right approach for several reasons:

If you're still interested, we can merge this! Let me know...

ewlsh commented:
> After further consideration and discussion (with a friend of mine), I've changed my position on this feature.
> 
> Initially, I was hesitant about adding `TsSatisfiesExpression` support directly to the core, thinking it might be better as a plugin. However, I now believe this is the right approach for several reasons:
> 
> If you're still interested, we can merge this! Let me know...

@adrai i've updated it and cleaned up the code, would happily have this as part of core.

adrai commented:
included in v1.9.0
